### PR TITLE
Support scalar mul of arkworks curve using precomputed table

### PR DESCRIPTION
**Pull request by tsunrise**
Adds support for scalar multiplication using precomputed bases. 

Closes: #24

### CODE REVIEW

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:432]: The `N` here represents the compile-time fixed and known number of bits of a scalar for the particular curve. This trait should not be implemented over all `N` (I wanted to use an associated constant but it's unstable to use that with slices rn). We can improve the documentation to help make this clearer.

File: manta-crypto/src/ecc.rs
Comment (bhgomes) [inline manta-crypto/src/ecc.rs:160]: Because the trait should be implemented for a single value of `N`, this should not be deleted.

**Review by bhgomes — CHANGES_REQUESTED** [review]

File: manta-crypto/src/ecc.rs
Comment (tsunrise) [reply manta-crypto/src/ecc.rs:160]: Implementation of `1` conflicts with implementation of `N`. Any good way to bypass this?

Comment (tsunrise) [reply manta-crypto/src/ecc.rs:160]: ah ok, I see the previous comment. so we are not implementing this for all `N`?

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:432]: To which specific value of `N` shall we implement? I can write a macro to do that

Comment (bhgomes) [reply manta-pay/src/crypto/ecc/arkworks.rs:432]: It should be implemented for the number of bits in a scalar. It's exactly the "debug assert" that you do, but we know at compile-time that the table is the correct size.

File: manta-crypto/src/ecc.rs
Comment (bhgomes) [reply manta-crypto/src/ecc.rs:160]: Yes, only for the exact `N` which is correct.

**Review by bhgomes — CHANGES_REQUESTED** [review]

File: manta-crypto/src/ecc.rs
Comment (tsunrise) [reply manta-crypto/src/ecc.rs:160]: Got it thanks!

Comment (tsunrise) [reply manta-crypto/src/ecc.rs:160]: Side note: currently we cannot put associate constant to const generic. For example, we cannot implement `ecc::PreprocessedScalarMul<Compiler<C>, Self::Scalar::MODULUS_BITS>`. We might need to implement that for every individual field.

Comment (tsunrise) [reply manta-crypto/src/ecc.rs]: I have implemented this trait only for Bls12_381. I also wrote a macro to easily support other curves if we need that

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: currently we cannot put associate constant to const generic. For example, we cannot implement `ecc::PreprocessedScalarMul<Compiler<C>, Self::Scalar::MODULUS_BITS>`. We might need to implement that for every individual curve. 

I have implemented this trait only for Bls12_381. I also wrote a macro to easily support other curves if we need that

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: fixed

**Comment by tsunrise** [comment]
Can someone allow force push on branches other than `main`? I messed up with commit histories @bhgomes @stechu

**Comment by tsunrise** [comment]
Can someone allow force push on branches other than `main`? I messed up with commit histories @bhgomes @stechu

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:445]: You can use the compile-time known size:

```rust
impl_processed_scalar_mul!(EdwardsProjective, {
    <<EdwardsProjective as ProjectiveCurve>::ScalarField as PrimeField>::Params::MODULUS_BITS
        as usize
});
```

We can incorporate this into the macro eventually to make it easier to implement.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:416]: Please `#[inline]` this method.

**Review by bhgomes — CHANGES_REQUESTED** [review]
Left some comments on the implementation that should be addressed before merging. I've also left some notes on some infrastructure we should add to improve development in ECC.

File: manta-crypto/src/ecc.rs
Comment (tsunrise) [reply manta-crypto/src/ecc.rs:142]: resolved offline

File: manta-crypto/src/ecc.rs
Comment (tsunrise) [inline manta-crypto/src/ecc.rs:138]: I will add some tests for this in a day or two. Do not merge until this comment is marked resolved

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: I agree. Removed.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: resolve first comment.  For second comment, see #39

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: removed

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: done

File: manta-crypto/src/ecc.rs
Comment (tsunrise) [reply manta-crypto/src/ecc.rs:138]: test added

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:39]: Move this non-public `use` import together with the other ones. The ordering of imports is

1. `use`
2. `pub use`
3. `mod`
4. `pub mod`

So non-`pub` before `pub`. Also, non-`feature` before `feature` gated and features in alphabetical order.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:367]: Please use `derivative::Derivative` for `derive` traits when there are type parameters on the deriving type so we have the correct semantics.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:552]: Do we really need this clone? Can we not do `Self::new(&self.0 + &rhs.0)`? I believe `+` already does `+=` internally so it won't have a performance penalty.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:582]: Please leave a newline between the imports and the first definition. Also, add a documentation string here.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs]: You can just do something like:

```rust
preprocessed_scalar_mul_test_template::<EdwardsProjective, EdwardsVar, { 
    ScalarParam<EdwardsProjective>::MODULUS_BITS
}>(&mut OsRng);
```

however it gets formatted, let `rustfmt` decide.

File: manta-pay/Cargo.toml
Comment (bhgomes) [inline manta-pay/Cargo.toml:81]: Instead of adding to the `test` feature, you should add this to the `dev-dependencies`. The `test` feature is for _exported_ test content rather than non-exported tests. NB: the `#[cfg(test)]` does not observe this feature only `#[cfg(feature = "test")]` does.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:419]: Can you try to make an alias for this? Not sure if it'll work because it may involve an unstable feature, but worth a try:

```rust
type ScalarParams<C> = <<C as ProjectiveCurve>::ScalarField as PrimeField>::Params;
```

Then you can call it as `{ ScalarParams::<$curve>::MODULUS_BITS }`. Also, we can call this from the test.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:550]: Please `#[inline]`. I'd like to inline all `impl Trait` methods for now.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:565]: Please `#[inline]`. I'd like to inline all `impl Trait` methods for now.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs]: Make sure to run `cargo fmt` on nightly to use the formating conventions the CI will accept.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:586]: Feel free to import `PreprocessedScalarMul` to avoid `ecc::` prefix. It's more important the local code is as simple as possible even if we have longer imports.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:609]: Feel free to use allocation strategies that reduce the amount of explicit type-annotations and trait names. Also, try to maximize the number of automatically deduced types by removing them one-by-one until either no types are specified or the compiler reaches an ambiguous typing state.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:612]: If you'd like, add an `is_satisfied` method to the `cs` directly (the `arkworks` concrete type not the trait definition).

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:602]: Unless you are also going to check something with the `base` and `scalar` directly, compress these groupings into

```rust
let base = Group::<C>::gen(rng).as_known::<Secret, _>(&mut cs);
```

(or whatever the minimally type-annotated form is) and likewise for `scalar`.

**Review by bhgomes — CHANGES_REQUESTED** [review]
Just a few more changes and we'll be good to go!

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:616]: You should leave off `test` prefixes and suffixes from the names of tests since `mod test` will already pick up a `test::` prefix.

**Review by bhgomes — CHANGES_REQUESTED** [review]
One more thing about test naming.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:552]: unfortunately `CurveVar` does not implement `Add` trait :( We can upstream this impl to arkworks in the future

File: manta-pay/Cargo.toml
Comment (tsunrise) [reply manta-pay/Cargo.toml:81]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:39]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:367]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:419]: it works! done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:550]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:565]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:582]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:586]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:602]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:609]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:612]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs:616]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:26]: I'm using the convention to inline this macro at the call site to reduce the number of imports, since this macro is used a lot, and importing this does not reduce the complexity at the call site:

```rust
#[derive(derivative::Derivative)]
```

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:30]: For these imports, use `self`:

```rust
ecc::{self, PointAdd, PointDouble}
```

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:576]: We use the name `test` for this module by convention. Unless specifically done so, all modules should be named in singular form rather than plural.

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (bhgomes) [inline manta-pay/src/crypto/ecc/arkworks.rs:616]: Please remove excess whitespace and use `base.scalar_mul(&scalar, &mut cs)` for `expected`.

**Review by bhgomes — CHANGES_REQUESTED** [review]

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: done

File: manta-pay/src/crypto/ecc/arkworks.rs
Comment (tsunrise) [reply manta-pay/src/crypto/ecc/arkworks.rs]: done

File: manta-crypto/src/ecc.rs
Comment (tsunrise) [reply manta-crypto/src/ecc.rs]: done

**Review by bhgomes — APPROVED** [review]

**Pull request merged**
